File: benchmarks/baselines/typescript/v2/service.ts
/** Transport-independent create-job handler. */

import {
  InsertOutcome,
  type CreateJobRequest,
  type CreateJobResult,
  type Job,
  type JobStore,
  type Priority,
  type ValidationIssue,
} from "./domain.js";

const jobIdPattern = /^[A-Za-z0-9][A-Za-z0-9_-]{0,63}$/u;
const controlCharacterPattern = /\p{Cc}/u;

export function createJob(
  request: CreateJobRequest,
  store: JobStore,
): CreateJobResult {
  const { priority, issues } = validate(request);
  if (issues.length > 0) {
    return { kind: "invalid", issues };
  }
  if (priority === undefined) {
    throw new Error("validated request must have a priority");
  }

  const job: Job = {
    jobId: request.jobId,
    task: request.task,
    payload: Uint8Array.from(request.payload),
    priority,
  };
  const outcome = store.insertIfAbsent(job);
  switch (outcome) {
    case InsertOutcome.inserted:
      return { kind: "created", job };
    case InsertOutcome.duplicate:
      return { kind: "already_exists", jobId: job.jobId };
    case InsertOutcome.unavailableBeforeCommit:
      return { kind: "persistence_unavailable" };
    default:
      return assertNever(outcome);
  }
}

function validate(request: CreateJobRequest): Readonly<{
  priority: Priority | undefined;
  issues: readonly ValidationIssue[];
}> {
  const issues: ValidationIssue[] = [];
  if (request.jobId.length === 0) {
    issues.push({ field: "job_id", reason: "missing" });
  } else if (!jobIdPattern.test(request.jobId)) {
    issues.push({ field: "job_id", reason: "invalid_format" });
  }

  if (request.task.length === 0) {
    issues.push({ field: "task", reason: "missing" });
    // Array.from counts Unicode code points, which are the fixture's scalar unit.
  } else if (Array.from(request.task).length > 80) {
    issues.push({ field: "task", reason: "too_long" });
  } else if (controlCharacterPattern.test(request.task)) {
    issues.push({ field: "task", reason: "control_character" });
  }

  if (request.payload.byteLength > 4096) {
    issues.push({ field: "payload", reason: "payload_too_large" });
  }
  if (request.priority === undefined) {
    issues.push({ field: "priority", reason: "missing" });
  }
  return { priority: request.priority, issues };
}

function assertNever(value: never): never {
  throw new Error(`invalid insert outcome: ${String(value)}`);
}

File: benchmarks/baselines/python/v2/service.py
"""Transport-independent create-job handler."""

from __future__ import annotations

import re
import unicodedata
from typing import assert_never

from v2.domain import (
    AlreadyExists,
    Created,
    CreateJobRequest,
    CreateJobResult,
    InsertOutcome,
    Invalid,
    Job,
    JobStore,
    PersistenceUnavailable,
    Priority,
    ValidationField,
    ValidationIssue,
    ValidationReason,
)

_JOB_ID_PATTERN = re.compile(r"[A-Za-z0-9][A-Za-z0-9_-]{0,63}\Z")


def create_job(request: CreateJobRequest, store: JobStore) -> CreateJobResult:
    """Complete validation before the only capability call."""
    priority, issues = _validate(request)
    if issues:
        return Invalid(issues)
    if priority is None:
        msg = "validated request must have a priority"
        raise AssertionError(msg)

    job = Job(request.job_id, request.task, request.payload, priority)
    outcome = store.insert_if_absent(job)
    match outcome:
        case InsertOutcome.INSERTED:
            return Created(job)
        case InsertOutcome.DUPLICATE:
            return AlreadyExists(job.job_id)
        case InsertOutcome.UNAVAILABLE_BEFORE_COMMIT:
            return PersistenceUnavailable()
        case _:
            assert_never(outcome)


def _validate(
    request: CreateJobRequest,
) -> tuple[Priority | None, tuple[ValidationIssue, ...]]:
    issues: list[ValidationIssue] = []

    if not request.job_id:
        issues.append(ValidationIssue(ValidationField.JOB_ID, ValidationReason.MISSING))
    elif _JOB_ID_PATTERN.fullmatch(request.job_id) is None:
        issues.append(
            ValidationIssue(
                ValidationField.JOB_ID,
                ValidationReason.INVALID_FORMAT,
            )
        )

    if not request.task:
        issues.append(ValidationIssue(ValidationField.TASK, ValidationReason.MISSING))
    elif len(request.task) > 80:
        issues.append(ValidationIssue(ValidationField.TASK, ValidationReason.TOO_LONG))
    elif any(unicodedata.category(char) == "Cc" for char in request.task):
        issues.append(
            ValidationIssue(
                ValidationField.TASK,
                ValidationReason.CONTROL_CHARACTER,
            )
        )

    if len(request.payload) > 4096:
        issues.append(
            ValidationIssue(
                ValidationField.PAYLOAD,
                ValidationReason.PAYLOAD_TOO_LARGE,
            )
        )
    if request.priority is None:
        issues.append(
            ValidationIssue(ValidationField.PRIORITY, ValidationReason.MISSING)
        )
    return request.priority, tuple(issues)
